mrmustard/physics/bargmann.py

```diff
@@ -103,3 +104,164 @@ def wigner_to_bargmann_U(X, d):
     N = X.shape[-1] // 2
     A, B, C = wigner_to_bargmann_Choi(X, math.zeros_like(X), d)
     return A[2 * N :, 2 * N :], B[2 * N :], math.sqrt(C)
+
+
+def complex_gaussian_integral(
+    Abc: tuple, idx_z: tuple[int, ...], idx_zconj: tuple[int, ...], measure: float = -1
+):
+    r"""Computes the Gaussian integral of the exponential of a complex quadratic form.
+    The integral is defined as (note that in general we integrate over a subset of 2m dimensions):
+
+    :math:`\int_{C^m} F(z) d\mu(z)`
+
+    where
+
+    :math:`F(z) = \textrm{exp}(-0.5 z^T A z + b^T z)`
+
+    Here, ``z`` is an ``n``-dim complex vector, ``A`` is an ``n x n`` complex matrix,
+    ``b`` is an ``n``-dim complex vector, ``c`` is a complex scalar, and :math:`d\mu(z)`
+    is a non-holomorphic complex measure over a subset of m pairs of z,z* variables. These
+    are specified by the indices ``idx_z`` and ``idx_zconj``. The ``measure`` parameter is
+    the exponent of the measure:
+
+    :math: `dmu(z) = \textrm{exp}(m * |z|^2) \frac{d^{2n}z}{\pi^n} = \frac{1}{\pi^n}\textrm{exp}(m * |z|^2) d\textrm{Re}(z) d\textrm{Im}(z)`
+
+    Arguments:
+        A,b,c: the ``(A,b,c)`` triple
+        idx_z: the tuple of indices of the z variables
+        idx_zconj: the tuple of indices of the z* variables
+        measure: the exponent of the measure (default is -1: Bargmann measure)
+
+    Returns:
+        The ``(A,b,c)`` triple of the result of the integral
+    """
+    A, b, c = Abc
+    if len(idx_z) != len(idx_zconj):
+        raise ValueError("idx_z and idx_zconj must have the same length")
+    n = len(idx_z)
+    idx = tuple(idx_z) + tuple(idx_zconj)
+    not_idx = tuple(i for i in range(A.shape[-1]) if i not in idx)
+
+    I = math.eye(n, dtype=A.dtype)
+    Z = math.zeros((n, n), dtype=A.dtype)
+    X = math.block([[Z, I], [I, Z]])
+    M = math.gather(math.gather(A, idx, axis=-1), idx, axis=-2) + X * measure
+    D = math.gather(math.gather(A, idx, axis=-1), not_idx, axis=-2)
+    R = math.gather(math.gather(A, not_idx, axis=-1), not_idx, axis=-2)
+
+    bM = math.gather(b, idx, axis=-1)
+    bR = math.gather(b, not_idx, axis=-1)
+
+    A_post = R - math.matmul(D, math.inv(M), math.transpose(D))
+    b_post = bR - math.matvec(D, math.solve(M, bM))
+    c_post = (
+        c * math.sqrt((-1) ** n / math.det(M)) * math.exp(-0.5 * math.sum(bM * math.solve(M, bM)))
+    )
+
+    return A_post, b_post, c_post
+
+
+def real_gaussian_integral(
+    Abc: tuple, idx: tuple[int, ...], measure: float = 0
+):  # pragma: no cover (will test it when we will use it)
```

Comment:
How do we know it works? Either add a test or remove it, and we will add it when we use it

Reply:
An integration test? 😜
Yeah I'll remove it